tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -52,6 +52,10 @@ using namespace at::indexing;
 */
 
 TEST_F(PipelineTest, Pipeline) {
+  if (communicator->size() < 6 || torch::cuda::device_count() < 6) {
```

Comment:
Should `communicator->size()` and `torch::cuda::device_count()` be the same? Iow, what would break if I say 

```
const auto device_count = torch::cuda::device_count();
NVF_ERROR(communicator->size() == device_count);
if (device_count < 6) {
```

Reply:
It should be `communicator->size() <=  torch::cuda::device_count()` since it's fine to use less than the available number of devices. I'll add this check the `MutiDeviceEnvinroment::SetUp()` since if we try to create a communicator thats greater than the number of devices everything breaks.  Good catch!

Reply:
It is not the same because communicator->size is the number of process spwaned, not number of GPU. Also (to be confirmed) torch::cuda::device_count() only counts the devices on the current node

Reply:
oh I forgot device_count() only returns the number of devices on the current node. I will not add that check then.

Reply:
I think it is not necessary to add those tests, which are all already included in `MultiDeviceExecutor::validate()`. This function is called inside `PipelineTest::execute()` -- the only missing part is to propagate the skip outside the function to skip executing `PipelineTest::validate()`

does it make sens? :)

Reply:
> the only missing part is to propagate the skip outside the function

$0.02: to simplify, it may even be possible to GTEST_SKIP inside execute(), although I haven't tried this and [the doc](https://google.github.io/googletest/advanced.html#skipping-test-execution) is unclear.

Reply:
I don't think this is possible. Since `execute()` is calling GTEST macros it has to be a void function, so we can't return the error message from `execute` back to the test.

Reply:
> I don't think this is possible. Since `execute()` is calling GTEST macros it has to be a void function, so we can't return the error message from `execute` back to the test.

I would use a boolean class data member

Reply:
We could do that, but imo it's cleaner to perform the check in the test rather than add a new class field.

Reply:
The check should be perform in all the tests, not only this one. So in my opinion it is more convenient to have it automatically done at one place instead of duplicating it everywhere with hardcoded values, even if it is at the cost of one supplmentary class member.

Another good solution would be to merge `execute` and `validate` into one function -- they are always used together for now anyway, and it would save us the method executeAndValidate

Reply:
Unfortunately, GTEST_SKIP only skips tests when called in either the enivronment's SetUp or the unit test itself, so combining execute and validate into one function and calling GTEST_SKIP will not skip the test. 

I can save the error message from validate and check it's results after each call executeValidate, but we will have to make certain to check it after executeValidate in each test otherwise the message won't get set.

Reply:
> I can save the error message from validate and check it's results after each call executeValidate, but we will have to make certain to check it after executeValidate in each test otherwise the message won't get set.

We could just let execute or executeValidate return the error message (or "") to the caller (i.e. the test case itself), which is then responsible to call `GTEST_SKIP`. Would that work? 

That being said, if easy, I prefer GTEST_SKIP up front in the test so it's immediately clear to the reader why a test is skipped. Currently, it's unclear to the reader (like me) why MultiDeviceExecutor::validate would fail. If the skip condition is indeed as simple as device_count<6, GTEST_SKIP that up front brings lots of clarity.

Reply:
Resolved offline with Sam. GTEST_SKIP is still in the helper function `executeAndValidate` and working properly now.